packages/grafana-ui/src/components/TransformersUI/TransformationsEditor.tsx

```diff
@@ -0,0 +1,127 @@
+import { DataTransformerID, DataTransformerConfig, DataFrame, transformDataFrame } from '@grafana/data';
+import { Select } from '../Select/Select';
+import { transformersUIRegistry } from './transformers';
+import React from 'react';
+import { TransformationRow } from './TransformationRow';
+import { Button } from '../Button/Button';
+import { css } from 'emotion';
+
+interface TransformationsEditorState {
+  updateCounter: number;
+}
+
+interface TransformationsEditorProps {
+  onChange: (transformations: DataTransformerConfig[]) => void;
+  transformations: DataTransformerConfig[];
+  getCurrentData: (applyTransformations?: boolean) => DataFrame[];
+}
+
+export class TransformationsEditor extends React.PureComponent<TransformationsEditorProps, TransformationsEditorState> {
+  state = { updateCounter: 0 };
+
+  onTransformationAdd = () => {
+    const { transformations, onChange } = this.props;
+    onChange([
+      ...transformations,
+      {
+        id: DataTransformerID.noop,
+        options: {},
+      },
+    ]);
+    this.setState({ updateCounter: this.state.updateCounter + 1 });
+  };
+
+  onTransformationChange = (idx: number, config: DataTransformerConfig) => {
+    const { transformations, onChange } = this.props;
+    transformations[idx] = config;
+    onChange(transformations);
+    this.setState({ updateCounter: this.state.updateCounter + 1 });
+  };
+
+  onTransformationRemove = (idx: number) => {
+    const { transformations, onChange } = this.props;
+    transformations.splice(idx, 1);
+    onChange(transformations);
+    this.setState({ updateCounter: this.state.updateCounter + 1 });
+  };
+
+  renderTransformationEditors = () => {
+    const { transformations, getCurrentData } = this.props;
+    const hasTransformations = transformations.length > 0;
+    const preTransformData = getCurrentData(false);
+
+    if (!hasTransformations) {
+      return undefined;
+    }
+
+    const availableTransformers = transformersUIRegistry.list().map(t => {
+      return {
+        value: t.transformer.id,
+        label: t.transformer.name,
+      };
+    });
+
+    return (
+      <>
+        {transformations.map((t, i) => {
+          let editor, input;
+          if (t.id === DataTransformerID.noop) {
+            return (
+              <Select
+                className={css`
+                  margin-bottom: 10px;
+                `}
+                key={`${t.id}-${i}`}
+                options={availableTransformers}
+                placeholder="Select transformation"
+                onChange={v => {
+                  this.onTransformationChange(i, {
+                    id: v.value as string,
+                    options: {},
+                  });
+                }}
+              />
+            );
+          }
+          const transformationUI = transformersUIRegistry.getIfExists(t.id);
+          input = transformDataFrame(transformations.slice(0, i), preTransformData);
```

Comment:
it looks like all transforms are passed the same preTransformData, how is this chained? and it this to support the debug mode?

Reply:
`preTransformData` is what query returned. We pass partially applied transformations to the next editors. This is/will be necessary for transformation editors (i.e. filter) to have access to field names/types that are available in the context of current transformation.